File: shared/domain/src/commonMain/kotlin/com/tddworks/claudebar/domain/monitor/QuotaMonitor.kt
package com.tddworks.claudebar.domain.monitor

import com.tddworks.claudebar.domain.model.QuotaStatus
import com.tddworks.claudebar.domain.model.UsageQuota
import com.tddworks.claudebar.domain.model.UsageSnapshot
import com.tddworks.claudebar.domain.provider.AIProvider
import com.tddworks.claudebar.domain.provider.AIProviderRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

/**
 * The main domain service that coordinates quota monitoring across AI providers.
 * Providers are rich domain models that own their own snapshots.
 * QuotaMonitor coordinates refreshes and alerts users when status changes.
 */
class QuotaMonitor(
    private val providers: AIProviderRepository,
    private val alerter: QuotaAlerter? = null
) {
    /** Previous status for change detection */
    private val previousStatuses = mutableMapOf<String, QuotaStatus>()

    /** Current monitoring job */
    private var monitoringJob: Job? = null

    /** Whether monitoring is active */
    private val _isMonitoring = MutableStateFlow(false)
    val isMonitoring: StateFlow<Boolean> = _isMonitoring.asStateFlow()

    /** The currently selected provider ID (for UI display) */
    private val _selectedProviderId = MutableStateFlow("claude")
    val selectedProviderId: StateFlow<String> = _selectedProviderId.asStateFlow()

    init {
        selectFirstEnabledIfNeeded()
    }

    // MARK: - Monitoring Operations

    /**
     * Refreshes all enabled providers concurrently.
     * Each provider updates its own snapshot.
     * Disabled providers are skipped.
     */
    suspend fun refreshAll() {
        coroutineScope {
            providers.enabled.forEach { provider ->
                launch {
                    refreshProvider(provider)
                }
            }
        }
    }

    /** Refreshes a single provider */
    private suspend fun refreshProvider(provider: AIProvider) {
        if (!provider.isAvailable()) {
            return
        }

        try {
            val snapshot = provider.refresh()
            handleSnapshotUpdate(provider, snapshot)
        } catch (e: Throwable) {
            // Provider stores error in lastError - no need for external observer
        }
    }

    /** Handles snapshot update and alerts user if status changed */
    private suspend fun handleSnapshotUpdate(provider: AIProvider, snapshot: UsageSnapshot) {
        val previousStatus = previousStatuses[provider.id] ?: QuotaStatus.HEALTHY
        val newStatus = snapshot.overallStatus

        previousStatuses[provider.id] = newStatus

        // Alert user only if status changed
        if (previousStatus != newStatus) {
            alerter?.alert(
                providerId = provider.id,
                previousStatus = previousStatus,
                currentStatus = newStatus
            )
        }
    }

    /** Refreshes a single provider by its ID */
    suspend fun refresh(providerId: String) {
        val provider = providers.provider(providerId) ?: return
        refreshProvider(provider)
    }

    /** Refreshes all enabled providers except the specified one */
    suspend fun refreshOthers(exceptProviderId: String) {
        val otherProviders = providers.enabled.filter { it.id != exceptProviderId }

        coroutineScope {
            otherProviders.forEach { provider ->
                launch {
                    refreshProvider(provider)
                }
            }
        }
    }

    // MARK: - Queries

    /** Returns the provider with the given ID */
    fun provider(id: String): AIProvider? = providers.provider(id)

    /** Returns all providers */
    val allProviders: List<AIProvider>
        get() = providers.all

    /** Returns only enabled providers */
    val enabledProviders: List<AIProvider>
        get() = providers.enabled

    /** Adds a provider dynamically */
    fun addProvider(provider: AIProvider) {
        providers.add(provider)
    }

    /** Removes a provider by ID */
    fun removeProvider(id: String) {
        providers.remove(id)
    }

    /** Returns the lowest quota across all enabled providers */
    fun lowestQuota(): UsageQuota? =
        providers.enabled
            .mapNotNull { it.snapshot.value?.lowestQuota }
            .minByOrNull { it.percentRemaining }

    /** Returns the overall status across enabled providers (worst status wins) */
    val overallStatus: QuotaStatus
        get() = providers.enabled
            .mapNotNull { it.snapshot.value?.overallStatus }
            .maxByOrNull { it.severity } ?: QuotaStatus.HEALTHY

    // MARK: - Selection

    /** The currently selected provider (from enabled providers) */
    val selectedProvider: AIProvider?
        get() = providers.enabled.find { it.id == _selectedProviderId.value }

    /** Status of the currently selected provider (for menu bar icon) */
    val selectedProviderStatus: QuotaStatus
        get() = selectedProvider?.snapshot?.value?.overallStatus ?: QuotaStatus.HEALTHY

    /** Whether any provider is currently refreshing */
    val isRefreshing: Boolean
        get() = providers.all.any { it.isSyncing.value }

    /** Selects a provider by ID (must be enabled) */
    fun selectProvider(id: String) {
        if (providers.enabled.any { it.id == id }) {
            _selectedProviderId.value = id
        }
    }

    /**
     * Sets a provider's enabled state.
     * When disabling the currently selected provider, automatically switches
     * to the first available enabled provider.
     */
    fun setProviderEnabled(id: String, enabled: Boolean) {
        val provider = providers.provider(id) ?: return
        provider.isEnabled = enabled
        if (!enabled) {
            selectFirstEnabledIfNeeded()
        }
    }

    /**
     * Selects the first enabled provider if current selection is invalid.
     * Called automatically during initialization and when providers are disabled.
     */
    private fun selectFirstEnabledIfNeeded() {
        if (providers.enabled.none { it.id == _selectedProviderId.value }) {
            providers.enabled.firstOrNull()?.let {
                _selectedProviderId.value = it.id
            }
        }
    }

    // MARK: - Continuous Monitoring

    /**
     * Starts continuous monitoring at the specified interval.
     * Returns a Flow of monitoring events.
     */
    fun startMonitoring(interval: Duration = 60.seconds): Flow<MonitoringEvent> {
        // Stop any existing monitoring
        monitoringJob?.cancel()

        _isMonitoring.value = true

        return flow {
            while (true) {
                refreshAll()
                emit(MonitoringEvent.Refreshed)
                delay(interval)
            }
        }
    }

    /** Stops continuous monitoring */
    fun stopMonitoring() {
        _isMonitoring.value = false
        monitoringJob?.cancel()
        monitoringJob = null
    }
}

File: Sources/Domain/Monitor/QuotaMonitor.swift
import Foundation
import Observation

/// Events emitted during continuous monitoring
public enum MonitoringEvent: Sendable {
    /// A refresh cycle completed
    case refreshed
    /// An error occurred during refresh for a provider
    case error(providerId: String, Error)
}

/// The main domain service that coordinates quota monitoring across AI providers.
/// Providers are rich domain models that own their own snapshots.
/// QuotaMonitor coordinates refreshes and alerts users when status changes.
@Observable
public final class QuotaMonitor: @unchecked Sendable {
    /// The providers repository (internal - access via delegation methods)
    private let providers: any AIProviderRepository

    /// Optional alerter for quota changes (e.g., system notifications)
    private let alerter: (any QuotaAlerter)?

    /// Clock for scheduling intervals (injectable for tests)
    private let clock: any Clock

    /// Previous status for change detection
    private var previousStatuses: [String: QuotaStatus] = [:]

    /// Current monitoring task
    private var monitoringTask: Task<Void, Never>?

    /// Whether monitoring is active
    public private(set) var isMonitoring: Bool = false

    /// The currently selected provider ID (for UI display)
    public var selectedProviderId: String = "claude"

    // MARK: - Initialization

    /// Creates a QuotaMonitor with a provider repository.
    /// Automatically validates the selected provider on initialization.
    public init(
        providers: any AIProviderRepository,
        alerter: (any QuotaAlerter)? = nil,
        clock: any Clock
    ) {
        self.providers = providers
        self.alerter = alerter
        self.clock = clock
        selectFirstEnabledIfNeeded()
    }

    // MARK: - Monitoring Operations

    /// Refreshes all enabled providers concurrently.
    /// Each provider updates its own snapshot.
    /// Disabled providers are skipped.
    public func refreshAll() async {
        await withTaskGroup(of: Void.self) { group in
            for provider in providers.enabled {
                group.addTask {
                    await self.refreshProvider(provider)
                }
            }
        }
    }

    /// Refreshes a single provider
    private func refreshProvider(_ provider: any AIProvider) async {
        guard await provider.isAvailable() else {
            return
        }

        do {
            let snapshot = try await provider.refresh()
            await handleSnapshotUpdate(provider: provider, snapshot: snapshot)
        } catch {
            // Provider stores error in lastError - no need for external observer
        }
    }

    /// Handles snapshot update and alerts user if status changed
    private func handleSnapshotUpdate(provider: any AIProvider, snapshot: UsageSnapshot) async {
        let previousStatus = previousStatuses[provider.id] ?? .healthy
        let newStatus = snapshot.overallStatus

        previousStatuses[provider.id] = newStatus

        // Alert user only if status changed
        if previousStatus != newStatus, let alerter = alerter {
            await alerter.alert(
                providerId: provider.id,
                previousStatus: previousStatus,
                currentStatus: newStatus
            )
        }
    }

    /// Refreshes a single provider by its ID.
    public func refresh(providerId: String) async {
        guard let provider = providers.provider(id: providerId) else {
            return
        }
        await refreshProvider(provider)
    }

    /// Refreshes the given providers once, preserving order and removing duplicates.
    public func refresh(providerIds: [String]) async {
        var seen = Set<String>()
        let uniqueProviderIds = providerIds.filter { providerId in
            seen.insert(providerId).inserted
        }

        await withTaskGroup(of: Void.self) { group in
            for providerId in uniqueProviderIds {
                group.addTask {
                    await self.refresh(providerId: providerId)
                }
            }
        }
    }

    /// Refreshes all enabled providers except the specified one.
    public func refreshOthers(except providerId: String) async {
        let otherProviders = providers.enabled.filter { $0.id != providerId }

        await withTaskGroup(of: Void.self) { group in
            for provider in otherProviders {
                group.addTask {
                    await self.refreshProvider(provider)
                }
            }
        }
    }

    // MARK: - Queries

    /// Returns the provider with the given ID
    public func provider(for id: String) -> (any AIProvider)? {
        providers.provider(id: id)
    }

    /// Returns all providers
    public var allProviders: [any AIProvider] {
        providers.all
    }

    /// Returns only enabled providers
    public var enabledProviders: [any AIProvider] {
        providers.enabled
    }

    /// Adds a provider dynamically
    public func addProvider(_ provider: any AIProvider) {
        providers.add(provider)
    }

    /// Removes a provider by ID
    public func removeProvider(id: String) {
        providers.remove(id: id)
    }

    /// Returns the lowest quota across all enabled providers
    public func lowestQuota() -> UsageQuota? {
        providers.enabled
            .compactMap(\.snapshot?.lowestQuota)
            .min()
    }

    /// Returns the selected quota for a provider from enabled provider snapshots.
    public func quota(providerId: String, quotaKey: String) -> UsageQuota? {
        providers.enabled
            .first { $0.id == providerId }?
            .snapshot?
            .quota(forKey: quotaKey)
    }

    /// Returns the menu bar percentage display for a provider/quota selection.
    public func menuBarPercentageDisplay(
        providerId: String,
        quotaKey: String,
        mode: UsageDisplayMode,
        burnRateWarningEnabled: Bool = false,
        burnRateThreshold: Double = 1.5
    ) -> MenuBarPercentageDisplay? {
        guard let quota = quota(providerId: providerId, quotaKey: quotaKey) else {
            return nil
        }

        return MenuBarPercentageDisplay(
            quota: quota,
            mode: mode,
            burnRateWarningEnabled: burnRateWarningEnabled,
            burnRateThreshold: burnRateThreshold
        )
    }

    /// Returns the menu bar duration display for a provider/quota selection.
    /// Sibling to `menuBarPercentageDisplay`; both use the same selectors.
    public func menuBarDurationDisplay(
        providerId: String,
        quotaKey: String,
        burnRateWarningEnabled: Bool = false,
        burnRateThreshold: Double = 1.5
    ) -> MenuBarDurationDisplay? {
        guard let quota = quota(providerId: providerId, quotaKey: quotaKey) else {
            return nil
        }

        return MenuBarDurationDisplay(
            quota: quota,
            burnRateWarningEnabled: burnRateWarningEnabled,
            burnRateThreshold: burnRateThreshold
        )
    }

    /// Builds the fully composed menu bar label for one or two quota windows.
    ///
    /// The primary window renders exactly as the single-window label always has
    /// (percentage and/or duration joined by " · "). When `secondaryQuotaKey` is
    /// non-empty and differs from the primary, a second window is appended: each
    /// window is prefixed with its `QuotaType.shortLabel` and the two are joined
    /// by " | ", e.g. "5h 12% | 7d 34%". The status is the most severe of the
    /// shown windows.
    ///
    /// Returns nil when neither percentage nor duration is enabled, or when no
    /// quota data is available for the requested windows.
    public func menuBarLabel(
        providerId: String,
        primaryQuotaKey: String,
        secondaryQuotaKey: String = "",
        showPercentage: Bool,
        showDuration: Bool,
        mode: UsageDisplayMode,
        burnRateWarningEnabled: Bool = false,
        burnRateThreshold: Double = 1.5
    ) -> MenuBarLabel? {
        func segment(forQuotaKey quotaKey: String) -> (text: String, status: QuotaStatus)? {
            let percentage = showPercentage
                ? menuBarPercentageDisplay(
                    providerId: providerId,
                    quotaKey: quotaKey,
                    mode: mode,
                    burnRateWarningEnabled: burnRateWarningEnabled,
                    burnRateThreshold: burnRateThreshold
                )
                : nil
            let duration = showDuration
                ? menuBarDurationDisplay(
                    providerId: providerId,
                    quotaKey: quotaKey,
                    burnRateWarningEnabled: burnRateWarningEnabled,
                    burnRateThreshold: burnRateThreshold
                )
                : nil

            switch (percentage, duration) {
            case let (.some(percentage), .some(duration)):
                return ("\(percentage.text) · \(duration.text)", percentage.status)
            case let (.some(percentage), .none):
                return (percentage.text, percentage.status)
            case let (.none, .some(duration)):
                return (duration.text, duration.status)
            case (.none, .none):
                return nil
            }
        }

        let primary = segment(forQuotaKey: primaryQuotaKey)
        let secondary = (!secondaryQuotaKey.isEmpty && secondaryQuotaKey != primaryQuotaKey)
            ? segment(forQuotaKey: secondaryQuotaKey)
            : nil

        switch (primary, secondary) {
        case let (.some(primary), .some(secondary)):
            let primaryLabel = QuotaType(quotaKey: primaryQuotaKey)?.shortLabel ?? primaryQuotaKey
            let secondaryLabel = QuotaType(quotaKey: secondaryQuotaKey)?.shortLabel ?? secondaryQuotaKey
            return MenuBarLabel(
                text: "\(primaryLabel) \(primary.text) | \(secondaryLabel) \(secondary.text)",
                status: max(primary.status, secondary.status)
            )
        case let (.some(primary), .none):
            return MenuBarLabel(text: primary.text, status: primary.status)
        case let (.none, .some(secondary)):
            return MenuBarLabel(text: secondary.text, status: secondary.status)
        case (.none, .none):
            return nil
        }
    }

    /// Returns the overall status across enabled providers (worst status wins)
    public var overallStatus: QuotaStatus {
        providers.enabled
            .compactMap(\.snapshot?.overallStatus)
            .max() ?? .healthy
    }

    // MARK: - Selection

    /// The currently selected provider (from enabled providers)
    public var selectedProvider: (any AIProvider)? {
        providers.enabled.first { $0.id == selectedProviderId }
    }

    /// Status of the currently selected provider (for menu bar icon)
    public var selectedProviderStatus: QuotaStatus {
        selectedProvider?.snapshot?.overallStatus ?? .healthy
    }

    /// Whether any provider is currently refreshing
    public var isRefreshing: Bool {
        providers.all.contains { $0.isSyncing }
    }

    /// Selects a provider by ID (must be enabled)
    public func selectProvider(id: String) {
        if providers.enabled.contains(where: { $0.id == id }) {
            selectedProviderId = id
        }
    }

    /// Sets a provider's enabled state.
    /// When disabling the currently selected provider, automatically switches
    /// to the first available enabled provider.
    public func setProviderEnabled(_ id: String, enabled: Bool) {
        guard let provider = providers.provider(id: id) else { return }
        provider.isEnabled = enabled
        if !enabled {
            selectFirstEnabledIfNeeded()
        }
    }

    /// Selects the first enabled provider if current selection is invalid.
    /// Called automatically during initialization and when providers are disabled.
    private func selectFirstEnabledIfNeeded() {
        if !providers.enabled.contains(where: { $0.id == selectedProviderId }),
           let firstEnabled = providers.enabled.first {
            selectedProviderId = firstEnabled.id
        }
    }

    // MARK: - Continuous Monitoring

    /// Refreshes only the currently selected provider.
    public func refreshSelected() async {
        await refresh(providerId: selectedProviderId)
    }

    /// Starts continuous monitoring at the specified interval.
    /// By default, refreshes the currently selected provider each cycle to minimize energy usage.
    /// When provider IDs are supplied, refreshes that de-duplicated provider set each cycle.
    /// Returns an AsyncStream of monitoring events.
    public func startMonitoring(
        interval: Duration = .seconds(60),
        providerIds: [String]? = nil
    ) -> AsyncStream<MonitoringEvent> {
        // Stop any existing monitoring
        monitoringTask?.cancel()

        isMonitoring = true

        return AsyncStream { continuation in
            let task = Task {
                while !Task.isCancelled {
                    if let providerIds {
                        await self.refresh(providerIds: providerIds)
                    } else {
                        await self.refreshSelected()
                    }
                    continuation.yield(.refreshed)

                    do {
                        try await clock.sleep(for: interval)
                    } catch {
                        break
                    }
                }
                continuation.finish()
            }

            self.monitoringTask = task

            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }

    /// Stops continuous monitoring
    public func stopMonitoring() {
        isMonitoring = false
        monitoringTask?.cancel()
        monitoringTask = nil
    }
}
